### ISSUE TITLE
[Feature request] Offer completion when defining a variable that is already used but squiggled

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#24826]
I've run into this situation a few times, where I use a non-existing variable first, and then want to define it. When typing that definition, IntelliSense doesn't help typing the same identifier.

![image](https://user-images.githubusercontent.com/12466233/36189680-bf107864-1108-11e8-84d7-cfc8bde98881.png)

This could be extended to offer variable identifiers that already exist elsewhere in the code, since there is much re-use.

@CyrusNajmabadi Any thoughts?

I've not prototyped yet, but two implementation problems seem salient:
- currently, no completion is offered in such spot, rather than some soft completion. I don't know if that will be easy to relax without opening a dam and getting too many completions offered.
- I'm not sure how to efficiently list diagnostics within a certain scope, to collect squiggled identifiers. Any tips?

**Comment by CyrusNajmabadi** [comment]
I think this just goes into the category of "intelliSense should be able to complete any word written in your project anywhere".  There's no need for it to just be completing a missing identifier.  if you had written:

```c#
/// Use 'someIdentifier' to blah blah blah
bool M(string someI$$
```

We should be able to complete out to 'someIdentifier' here.

**Comment by Neme12** [comment]
@CyrusNajmabadi 
> intelliSense should be able to complete any word written in your project anywhere

I strongly disagree. I hate this kind of "completion" whenever I use VS Code and it always confuses me when I see something offered in the completion list that I think shouldn't be there. It always throws me off and makes me question my knowledge of the code. If I type `A.` and see 'B' offered I'm thinking "wait... why do I see B here? it shouldn't be in A... should it? is it?" and makes me lookup the definition only to actually see that I was correct and the completion was wrong.
Even if I got used to this over time, it would still make me trust completion a lot less, thus rely on it a lot less and thus be a lot less productive. And I'm not even talking about accidental commits of these "suggestions". (As far as I know, suggestion mode is a property that applies to the whole completion list and there's no way to only make some items soft select). But even if they did, I don't want to see this kind of nonsense next to regular IntelliSense. For me, completion is *the* source of truth. I don't want to see random words just because they appear anywhere in the file. That defeats the whole purpose.

Putting that aside, as far as the more specific feature about used but undeclared identifiers/members, I assume this would be symbol based as opposed to blindly showing list of all words in the file. Therefore I would expect this behavior:
```c#
Console.WriteLine(identifier);
$$ // I do see 'identifier' here
Console.$$ // I don't see 'identifier' here
```
```c#
class C
{
    int a;
    void M()
    {
        string a;
        this.a.identifier();
        this.a.$$ // 'identifier' offered here
        a.$$ // not offered here
        int b;
        b.$$ // offered here
    }
}
```
I could see how this could be useful and I'd be OK with this feature as long as it's sufficiently differentiated from regular completion, maybe even by showing an error squiggle for the item.

**Comment by Neme12** [comment]
But if this is only about suggested names during a declaration, then I don't mind unintelligent suggestions.

**Comment by jcouv** [comment]
> But if this is only about suggested names during a declaration, ...

Yes, my proposal was only to help type names in declarations.
Still, I'd prefer to use a narrower list of suggestions (looking for missing identifiers diagnostics in context), at least for an initial prototype. 
I need to find a way to collect relevant diagnostics (maybe from the current method, or file?).

**Comment by CyrusNajmabadi** [comment]
> I strongly disagree. I hate this kind of "completion" whenever I use VS Code and it always confuses me when I see something offered in the completion list that I think shouldn't be there. 

I didn't say it needed to be in the normal completion list.  I just said that "intelliSense should be able to complete any word written in your project anywhere".  In other words, you *as hte user* should be able to say something like "complete out this word i know was written somewhere, but which may not be a symbol name yet".

**Comment by genlu** [comment]
@CyrusNajmabadi This scenario is well covered by Copilot completion. And it seems you and @jcouv are suggesting different approach to address this?

**Comment by CyrusNajmabadi** [comment]
@genlu I'm fine with copilot being the solution here

**Comment by genlu** [comment]
Copilot works reasonably well picking up the undefined variable when the cursor is located at the end of the line, but it doesn't trigger when we are in the parameter list as shown in this issue.  There's discussion on improving this so I think it should work in future. 

![Image](https://github.com/user-attachments/assets/88e8f555-f577-45db-91e4-c84b53ffc6ed)

@jcouv are you Ok with this Copilot based approach? Otherwise, I do remember there's a way to get cached diagnostics (so it's fast enough to meet the perf requirement of completion) that maybe you could try?
https://github.com/dotnet/roslyn/blob/f28e37d6b41973d4d086904f8f2ed32ad8ca2002/src/LanguageServer/Protocol/Features/Diagnostics/DiagnosticAnalyzerService.cs#L101
